rustls/src/server/tls13.rs

```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
             .collect();
 
         let supported_groups = suites::KeyExchange::supported_groups();
-        let chosen_group = supported_groups
+
+        let chosen_share = supported_groups
             .iter()
-            .find(|group| share_groups.contains(group))
-            .cloned();
+            .find(|group| share_groups.contains(group)) // Find the chosen group
+            .and_then(|g| shares_ext.iter().find(|k| k.group == *g)); // Then its share
 
-        if chosen_group.is_none() {
-            // We don't have a suitable key share.  Choose a suitable group and
-            // send a HelloRetryRequest.
-            let retry_group_maybe = supported_groups
+        let chosen_share =  match chosen_share {
+            Some(s) => s,
+            None => {
+                // We don't have a suitable key share.  Choose a suitable group and
+                // send a HelloRetryRequest.
+                let retry_group_maybe = supported_groups
                 .iter()
                 .find(|group| groups_ext.contains(group))
                 .cloned();
```

Comment:
I think these will now need an extra bit of indentation.

Reply:
Good catch, thanks. Addressed.